src/pinghelper.cpp

```diff
@@ -31,53 +50,141 @@ void PingHelper::start(const QString& serverIpv4Gateway,
 
   m_gateway = serverIpv4Gateway;
   m_source = deviceIpv4Address.section('/', 0, 0);
+  m_pingSender =
+#if defined(MVPN_LINUX) || defined(MVPN_ANDROID)
+      new LinuxPingSender(m_source, this);
+#elif defined(MVPN_MACOS) || defined(MVPN_IOS)
+      new MacOSPingSender(m_source, this);
+#elif defined(MVPN_WINDOWS)
+      new WindowsPingSender(m_source, this);
+#else
+      new DummyPingSender(m_source, this);
+#endif
+  connect(m_pingSender, &PingSender::recvPing, this, &PingHelper::pingReceived);
+
+  // Reset the ping statistics
+  m_sequence = 0;
+  for (int i = 0; i < PING_STATS_WINDOW; i++) {
+    m_pingData[i].timestamp = -1;
+    m_pingData[i].latency = -1;
+    m_pingData[i].sequence = 0;
+  }
+
   m_pingTimer.start(PING_TIMOUT_SEC * 1000);
 }
 
 void PingHelper::stop() {
   logger.log() << "PingHelper deactivated";
 
+  if (m_pingSender) {
+    delete m_pingSender;
+    m_pingSender = nullptr;
+  }
+
   m_pingTimer.stop();
+}
 
-  for (PingSender* pingSender : m_pings) {
-    pingSender->deleteLater();
+void PingHelper::nextPing() {
+#ifdef QT_DEBUG
+  logger.log() << "Sending ping seq:" << m_sequence;
+#endif
+
+  int index = m_sequence % PING_STATS_WINDOW;
+  m_pingData[index].timestamp = QDateTime::currentMSecsSinceEpoch();
+  m_pingData[index].latency = -1;
+  m_pingData[index].sequence = m_sequence;
+  m_pingSender->sendPing(m_gateway, m_sequence);
+
+  m_sequence++;
+}
+
+void PingHelper::pingReceived(quint16 sequence) {
+  int index = sequence % PING_STATS_WINDOW;
+  if (m_pingData[index].sequence == sequence) {
+    qint64 sendTime = m_pingData[index].timestamp;
+    m_pingData[index].latency = QDateTime::currentMSecsSinceEpoch() - sendTime;
+    emit pingSentAndReceived(m_pingData[index].latency);
+#ifdef QT_DEBUG
+    logger.log() << "Ping answer received seq:" << sequence
+                 << "avg:" << latency()
+                 << "loss:" << QString("%1%").arg(loss() * 100.0)
+                 << "stddev:" << stddev();
+#endif
   }
-  m_pings.clear();
 }
 
-void PingHelper::nextPing() {
-  logger.log() << "Sending a new ping. Total:" << m_pings.length();
+uint PingHelper::latency() const {
+  int recvCount = 0;
+  qint64 totalMsec = 0;
 
-  PingSender* pingSender = new PingSender(this);
-  connect(pingSender, &PingSender::completed, this, &PingHelper::pingReceived);
-  m_pings.append(pingSender);
-  pingSender->send(m_gateway, m_source);
+  for (const PingSendData& data : m_pingData) {
+    if (data.latency < 0) {
+      continue;
+    }
+    recvCount++;
+    totalMsec += data.latency;
+  }
 
-  while (m_pings.length() > PINGS_MAX) {
-    m_pings.at(0)->deleteLater();
-    m_pings.removeAt(0);
+  if (recvCount <= 0) {
+    return 0.0;
   }
+
+  totalMsec += recvCount / 2;
```

Comment:
I don't understand this part. I thought you want to compute the average which should be: `totalMsec / recvCount`. Why do you add this extra `recvCount /2`.
Maybe write a comment

Reply:
This is a math trick that produced the nearest-integer rounding instead of the floor when doing an integer divide.